### PR TITLE
Add option for annotation replies

### PR DESCRIPTION
**Pull request by joemoongit**
Creating a draft pr for now. Still need to write test cases and may make modifications.

### CODE REVIEW

**Comment by julianhille** [comment]
Are you fine when i run ci after v4 release?

**Comment by joemoongit** [comment]
yes

**Comment by julianhille** [comment]
damn also needs rebasing to be able to run in CI

**Comment by joemoongit** [comment]
i’ve marked this pull request as ready. the annotation replies are working and they show up on pdfs. i've also updated how the annotations show up for the function text(). there are some more options we can add (i have commented them out for now) and optimizations we could make, but i will handle it in another pr, if that is okay with you. we could also add support for other types of annotations in later PRs.

**Comment by julianhille** [comment]
will try this evening or tomorrow to get back to it, please remind me if i do not do this. sorry for the inconveniences

**Comment by joemoongit** [comment]
do you have any suggestions? are you ok with the naming of some of the new variables such as 'replies' in AnnotOptions? also are you ok with me making some refactors/additions in later PRs?

File: lib/recipe/annotation.js [hunk 1]
```diff
@@ -59,8 +64,9 @@ exports.annot = function annot(
 // Link, Popup, Movie, Widget, Screen, PrinterMark, TrapNet, Watermark, 3D
 exports._attachNonMarkupAnnot = function _attachNonMarkupAnnot() {};
 
-exports._annot = function _annot(subtype, args = {}, pageNumber) {
-  const { x, y, width, height, text, options } = args;
+exports._annot = function _annot(subtype, args = {}, pageNumber, ref) {
```
Comment (julianhille) [inline lib/recipe/annotation.js:67]: has 4 function arguments but later you only call it with three, we should mark it opional. ahh ts is not supported. meh and its not public api

File: lib/recipe/annotation.js [hunk 2]
```diff
@@ -103,6 +109,19 @@ exports._annot = function _annot(subtype, args = {}, pageNumber) {
   const ey = nHeight ? nHeight : 0;
   const position = [nx, ny, nx + ex, ny + ey];
 
+  if (reply && ref) {
+    text = reply.text;
+    params.title = reply.title || params.title;
+    params.date = reply.date || params.date;
+    params.subject = reply.subject || params.subject;
+
+    // wip
```
Comment (julianhille) [inline lib/recipe/annotation.js]: still wip?

File: tests/recipe/annotation-text.js [hunk 3]
```diff
@@ -104,4 +104,197 @@ describe("Annotation: Text Annotations", () => {
...
+        date: "2023-07-11T06:28:42.090Z",
+        subject: "A subject 23",
+      })
+      .text("This text should be underlined with custom value.", 50, 250, {
+        underline: {
+          color: [255, 0, 128],
+          text: "Underline!",
+        },
+        size: 30,
+        title: "Guido van Rossum",
```
Comment (julianhille) [inline tests/recipe/annotation-text.js:238]: :>

File: lib/recipe/text.js [hunk 4]
```diff
@@ -238,6 +238,13 @@ exports._makeTextBox = function _makeTextBox(options) {
  * @param {string|number[]} [options.textBox.style.fill] - Text Box border background color (HexColor, PercentColor or DecimalColor)
  * @param {number} [options.textBox.style.opacity=1] - Text Box border background opacity
  * @param {boolean|number|number[]} [options.textBox.style.borderRadius=0] - Border radius to apply to get rounded corners.
+ * @param {string} [options.title] - Title of annotation
+ * @param {boolean} [options.open] - WIP
```
Comment (julianhille) [inline lib/recipe/text.js]: please update `WIP`

**Review by julianhille — CHANGES_REQUESTED** [review]
just a tiny bit to change overall good work thanks for the PR

**Comment by julianhille** [comment]
Do you want me to wait requested changes?

**Comment by joemoongit** [comment]
i will get back to you after this weekend. apologies for the delay

**Comment by joemoongit** [comment]
i've made some updates to the options for annotation replies. it appears that the 'open' and 'icon' keys only pertain to text annotations of which are not replies. i've updated the documentation and test cases as well

File: lib/recipe/annotation.js (hunk 2 above)
Comment (julianhille) [reply lib/recipe/annotation.js]: Wip comment

File: lib/recipe/annotation.js [hunk 5]
```diff
@@ -27,30 +27,35 @@ exports.comment = function comment(text = "", x, y, options = {}) {
...
+ * @param {string} subtype - The markup annotation type 'Text'|'Link'|'FreeText'|'Line'|'Square'|'Circle'|'Polygon'|'PolyLine'|'Highlight'|'Underline'|'Squiggly'|'StrikeOut'|'Caret'|'Stamp'|'Ink'|'Popup'|'FileAttachment'|'Sound'|'Movie'|'Screen'|'Widget'|'PrinterMark'|'TrapNet'|'Watermark'|'3D'|'Redact'|'Projection'|'RichMedia'
  * @param {Object} [options] - The options
  * @param {string} [options.title] - The title.
- * @param {boolean} [options.open=false] - Open the annotation by default?
+ * @param {boolean} [options.open=false] - Open the annotation. Annotation will be closed by default. Specific to text annotations; subtype='Text'
+ * @param {boolean} [options.richText] - Rich text
  * @param {'invisible'|'hidden'|'print'|'nozoom'|'norotate'|'noview'|'readonly'|'locked'|'togglenoview'} [options.flag] - The flag property
- * @param {'Comment'|'Key'|'Note'|'Help'|'NewParagraph'|'Paragraph'|'Insert'} [options.icon] - The icon of annotation.
+ * @param {'Comment'|'Key'|'Note'|'Help'|'NewParagraph'|'Paragraph'|'Insert'} [options.icon] - The icon of annotation. Specific to text annotations. Default value: 'Note'
```
Comment (julianhille) [inline lib/recipe/annotation.js]:
```suggestion
 * @param {'Comment'|'Key'|'Note'|'Help'|'NewParagraph'|'Paragraph'|'Insert'} [options.icon='Note'] - The icon of annotation. Specific to text annotations. Default value: 'Note'
```

**Review by julianhille — APPROVED** [review]

File: lib/recipe/annotation.js (hunk 2 above)
Comment (joemoongit) [reply lib/recipe/annotation.js]: no longer wip. annotation replies have the following properties: subtype, text, title, date, subject, richText, flag
latest commit does not have the comment

**Pull request merged**